binding/format/protobuf/v2/protobuf.go

```diff
@@ -106,19 +104,19 @@ func sdkToProto(e *event.Event) (*pb.CloudEvent, error) {
 	// NOTE: This is commented out until we add the data codec in a later patch
 	// or PR. Embedded here for illustration of how we will implement the
 	// requirement to use ProtoData IFF the data content type is protobuf.
-	// if e.DataContentType() == event.ApplicationProtobuf {
-	// 	anymsg := &anypb.Any{}
-	// 	if err := proto.Unmarshal(e.Data(), anymsg); err != nil {
-	// 		if e.DataSchema() == "" {
-	// 			return nil, fmt.Errorf("cannot encode direct protobuf message without dataschema. set dataschema to the appropriate protobuf type like type.googleapis.com/packge.v1.Type or make sure you are using the appropriate data content type %s", event.ApplicationProtobuf)
-	// 		}
-	// 		anymsg.TypeUrl = e.DataSchema()
-	// 		anymsg.Value = e.Data()
-	// 	}
-	// 	container.Data = &pb.CloudEvent_ProtoData{
-	// 		ProtoData: anymsg,
-	// 	}
-	// }
+	if e.DataContentType() == ContentTypeProtobuf {
+		anymsg := &anypb.Any{}
+		if err := proto.Unmarshal(e.Data(), anymsg); err != nil {
+			if e.DataSchema() == "" {
+				return nil, fmt.Errorf("cannot encode direct protobuf message without dataschema. set dataschema to the appropriate protobuf type like type.googleapis.com/packge.v1.Type or make sure you are using the appropriate data content type %s", ContentTypeProtobuf)
+			}
+			anymsg.TypeUrl = e.DataSchema()
+			anymsg.Value = e.Data()
+		}
```

Comment:
I want to highlight this section because it demonstrates the big question I have for this PR. There's a circular relationship between the codec and the format defined in the [protobuf spec](https://github.com/cloudevents/spec/blob/v1.0.1/protobuf-format.md) that is difficult to resolve without either changing the spec or introducing non-specified implementations details.

When both protobuf format and encoding are used together then we must use the `ProtoData` field of the protobuf CloudEvent type to store the data. I believe the intention of the protobuf specification is to match the behavior of JSON+JSON which embeds the data as a native type within the envelope rather than a JSON ecoded string within a JSON envelope. In the case of protobuf, though, the only generic type available is `Any` which contains two fields: a type URI and a protobuf encoded message.

In order to create an `Any` we must know the source type URI. The `Encode` implementation in this draft takes a source protobuf message and generates an `Any` from it such that the type URI is correct. However, this means the `Encode` and `Decode` methods practically only work correctly when used with the protobuf format. If used with the JSON format then `Encode` will incorrectly produce an encoded `Any` rather than an encoded form of the original message. Only users of this Go SDK would handle this case.

The problem I've highlighted here, though, is that there is not enough information within the `Event` during formatting to construct the `Any` required to comply with the spec. The special behavior of the format should be entirely encapsulated here but the type erasure that happens during the datacodec `Encode` leaves us without a type URI. The same type URI could be set as the dataschema value but this value is not required by either the primary CloudEvent spec or the protobuf extension.

Some options I've considered that I'd like feedback on:

1. Only convert from source data to `Any` in the format when protobuf content type is used. Use the dataschema value, including no value, as the type URI in the `Any`. Rely on the use of the non-type URI validating decoding feature of `Any` to prevent malformed or missing type URIs from blocking decoding back to a valid protobuf message.  This option technically results in an [invalid `Any`](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto#L125) but this would be hidden behind the format abstraction and never surfaced to users. The risk is that other languages protobuf implementations may not offer the option to bypass invalid or missing type URIs.

2. Change the protobuf specification to require the `dataschema` value and match it to the type URI that should be included in an `Any`. However, the type URI of a protobuf message is most often not a valid URI as required by the CloudEvent specification but a URI-reference because it is missing the required URI scheme. This means we would need to modify the actual type URI to something CloudEvent compliant in order to use that field. Additionally, the API for getting a message type URI is not a top-level feature but an artifact of the protobuf reflection interface in Go (ex: `string(msg.ProtoReflect().Descriptor().FullName())`). I'm not certain how other language implementation of protobuf expose this.

3. Artificially restrict the protobuf datacodec to only be used with the protobuf format. This would hide the issues without addressing them. However, this option also seems to require that we leak through the format/datacodec abstraction and add a special case somewhere in the main SDK that identifies and bails on this condition. Otherwise, it's not clear how we could impose that restriction only through the scope of this plugin.


I think Option 1 is the better of the batch because it best hides the implementation detail of using an `Any` and allows the various CloudEvent SDKs to manage the complexity of handling a technically invalid `Any` due to the potentially missing or malformed type URI. That puts the least burden on users of the SDK and works within the constraints of the current spec. That being said, I'm open to hearing more options or what folks think of the options I've listed.